### ISSUE TITLE
Sphinx-1.6 release plan

### ISSUE THREAD

**Issue by tk0miya** [sphinx-doc/sphinx#3586]
Hi maintainers,

According to our time based release policy, I'd propose you to release next major version in this June.
What do you think about the schedule?
I know there are still many issues for the release. So I might reconsider the schedule or the milestone of these issues.
https://github.com/sphinx-doc/sphinx/milestone/23

Personally, I feel the master branch has grown so much since last major release. We added many features and improved codebase much (We have many refactored and deprecated our codes!). So it would be nice to go forward with postponing some issues to next major release.

In addition, I'd like to propose not to release alpha packages from our process since this release. At last two major releases, we had provided alpha packages to request for comments for users and developers on early stage of our development.
But I feel that is not go well. Certainly, we'd received some issues for the release. But most of them were posted after final release. I think it would go well without alpha version because we still have beta release.
The alpha release also brings prolongation of release process to us.
For example, Sphinx-1.5 was released with following schedule:
* 1.5a1: 2016/9/21
* 1.5a2: 2016/10/17
* 1.5b1: 2016/11/6
* 1.5 (final): 2016/12/5
About three months were taken for 1.5 release (At 1.4, it took about 2 months).
So I think alpha release is not needed for us.

Please let me know your opinion.

**Comment by tk0miya** [comment]
My proposal (rough schedule)

# Plan A:
* Mid of May: feature freeze, call for translation and release 1.6a1
* End of May: release 1.6a2 (if needed)
* Mid of June: release 1.6 final

# Plan B:
* End of May: feature freeze, call for translation and release 1.6a1
* Mid of June: release 1.6a2 (if needed)
* End of June: release 1.6 final


.. note::

   We usually call for translation for collaborators during beta release.
   It would be shorten if we request before feature release.

**Comment by shimizukawa** [comment]
I agreed to release 1.6 in June.
And I also agreed to skip alpha.

BTW, I think 1.6a1 a2 in your both of plans must be 1.6b1 and b2.

**Comment by jfbu** [comment]
I too agree that alpha step could be skipped. in practice for LaTeX I feel there has been much refactoring  (particularly around tables, but there  is also new configuration for xelatex and lualatex and choice of fonts remains to be done) and I expect to anticipate feature freeze late April at latest so that in May I have some time to make final review with the hope not too many projects worldwide will be broken at 1.6 release ...  I know that already in June I will have little time, and that in July, August I basically have no internet. So I have slight preference for Plan A with its earlier scheduling, but this is not strong opinion.

**Comment by tk0miya** [comment]
Thank you for comment.

TL;DR

* We need to reconsider the plan because @jfbu will be absent after the release
* I propose Plan C (release in Mid of May) and Plan D (release in Sept)
* IMO, either is fine for me. So I feel Plan D is a bit better. 

Okay, your absent is absolutely most important schedule! (Personally, I believe OSS activity should not affect to real life!)
Based on that, we should reconsider the schedule because we introduced much changes into latex builder and its macros. I don't think they are breaking changes. If any bugs would found, we would need the @jfbu 's help. (Surely I'd reviewed the changes, but I have no confident to fix them...)

So I propose release plan again:

# Plan C:
* Apr. 22-23: Feature freeze, Call for translation and release 1.6b1
* May 13-14: release 1.6 final
* Until end of May: release 1.6.1 (bug fix release if needed)

In my experience, most of users reports their problems with new package after final release. So this plan shorten beta-cycle and hasten final release to keep time to fix the reported bugs in May or early June.

Pros:
* able to fix the bugs around LaTeX (if exists) in plenty of time
* able to drop supports of old codebase (There are too much difference between stable and master. The release resets the difference! If not released, the merging will become heavier and heavier...)
* able to drop supports of old environments (python, libraries, TeXLive and so on)

Cons:
* No time so much to add new more feature into 1.6 (Is this really cons? I believe the milestone label of issues is not a promise. Even if we release in June or July, we might not add all of features until the X-Day...)

# Plan D:

* Postpone the release to Sept. or Oct.

This plan simply postpones to the date when everybody fine. Nobody hurries us to the release. So far, Sphinx had never release the 6months cycle. It was usually taken 8-12mo. It's a trap of time based release policy.
v1.5 was released at 5/Dec/2016, so it is not too late if we'll ship it in Sept. or Oct.

Pros:
* Also be able to fix the bugs around LaTeX
* Many time to add the new feature to 1.6

Cons:
* We have to support old codebase and old environments
* The synchronizing cost between stable and master become larger and larger
* Nobody can't see future, so also can't promise we will be not in busy.


IMO, either is fine for me. So I feel Plan D is a bit better.

**Comment by shibukawa** [comment]
I like Plan C

**Comment by jfbu** [comment]
@tk0miya congratulations on very clear explanation of the pros and cons of Plans C and D. Both are fine with me.

The TeXLive release cycle is at http://www.tug.org/texlive/. If Sphinx 1.6 is released mid-May, we still will have had time to make test it with TeXLive 2017 (I will install the pretest of TeXLive 2017, as TL2016 will stop getting updates from CTAN at Apr'16, they will go only to the TL2017 pretest). I don't expect issues because Sphinx uses either old well-established LaTeX packages which do not change, or its own packages. As Sphinx contains no LuaTeX Lua code, there is no problem there to fear, and package luatex85 already handles TeXLive 2016 situation. Nevertheless it is a good if a minimal testing is done on pre-released TL2017 and the Plan C still leaves room for that.

In summary I have no strong opinion about Plan C or Plan D, but I am a bit tempted by Plan C because master branch contains many things which it would be nice to get released. Regarding LaTeX I hope requirement of use of latexmk for make latexpdf will not be too disruptive to people's installation... but I am really looking forward to see it part of official released Sphinx, because it speeds up my jobs with Sphinx latexpdf a lot.

**Comment by shimizukawa** [comment]
👍 to plan C 

There is large diff of master and stable already. I think it makes annoying us to develop

**Comment by tk0miya** [comment]
Okay, let's go with Plan C!
>Apr. 22-23: Feature freeze, Call for translation and release 1.6b1
>May 13-14: release 1.6 final

**Comment by jfbu** [comment]
Hi @tk0miya just to say that I won't be available for any Sphinx task starting today and until April 24. There a few issues assigned to me (https://github.com/sphinx-doc/sphinx/issues?q=is%3Aopen+assignee%3Ajfbu) which I thus will fix for 1.6b2 (or final 1.6 if no 1.6b2), not 1.6b1.

**Comment by tk0miya** [comment]
@jfbu No problem, all bug fixes are allowed after 1.6b1 (of course, after 1.6 final)!

**Comment by tk0miya** [comment]
Now #3562 is under reviewing. But it would be not ended until the feature freeze. So I'd like to merge it after the period. Any comments?

**Comment by tk0miya** [comment]
Hello, maintainers.

This is X-DAY for 1.6b1. Are you ready?
In Japan, I and @shimizukawa work for it in Sphinx hack-day event. So we try to clear (reduce?) the tasks marked as 1.6 milestone.
https://github.com/sphinx-doc/sphinx/milestone/23

If you have some troubles, please let me know. It might be accepted to postpone the release a few days.
If none, I will ship it in tonight (in JST; maybe noon - afternoon tea time in UTC)

**Comment by jfbu** [comment]
@tk0miya from my side I can't address #3479 in time for 1.6b1 because I am on tight schedule today and this is new feature which requires also coding in latex writer side. This would be new feature but on the other hand might be said bug-fix so perhaps for 1.6b2 ? Regarding #3532 and #3447, those would be bug-fixes and I can still do it after feature freeze. (but #3532 which is caused by threeparttable might be avoided by not using threeparttable, which might  be at 1.7; indeed I thought about it this morning (Europe time) and have only a half-fix when using threeparttable, so I can't contribute it for 1.6b1; and simplest of all would be not to use threeparttable at all for tables -- this is other discussion). 

Then #3562 we discussed already can be merged after 1.6b1 once it is resolved to complete satisfaction.

In summary, I have no more tasks pending which I can address in time for 1.6b1, so go ahead ;-)

**Comment by tk0miya** [comment]
@jfbu 1.6b1 is "feature" freeze, not "code" freeze. So you can continue to fix bugs until the release of 1.6-final. It means after the beta release, v1.6 is treated like stable branch.
#3479, #3532 and #3447 are already marked as "bug". So it's okay to fix them after 1.6b1 (of course, also okay for 1.7).

About #3562, I'll take a look later.

Thanks,

**Comment by tk0miya** [comment]
>About #3562, I'll take a look later.

I decided to merge it in next beta release (1.6b2?).
From now on, I start to packaging!

**Comment by tk0miya** [comment]
I released 1.6b1 just now!
https://pypi.python.org/pypi/Sphinx/1.6b1
I will send announce-mails tomorrow (I'm very tired and sleepy now. It's too late in Japan :-p)

@shimizukawa could you update i18n messages on transifex please? It is needed before call-for-translations.

**IMPORTANT**
From now on, the master branch is used for v1.7 development.
For v1.6 release, please use *1.6-release* branch.
As I commented above, 1.6-release branch behaves like stable branch. So please don't add new feature into the branch basically. If you need to add a new feature, please let us know. We'll discuss about it.

Note: As an exception, #3562 will be introduced soon.

Thank you for all your efforts!

**Comment by shimizukawa** [comment]
Thanks!
I'll make sure the translations.

2017年4月24日(月) 1:03 Takeshi KOMIYA <notifications@github.com>:

**Comment by shimizukawa** [comment]
sphinx translation catalog has been updated: ac5e76f5989eeb74b
This change contain a procedure to update sphinx.pot file.

**Comment by tk0miya** [comment]
Confirmed. Thank you for updating. To be enable to build our doc successfully on readthedocs, I will release 1.6b2 at this night. at same time, I will post call-for-translations announcement.
Thanks,

**Comment by tk0miya** [comment]
Now I released 1.6b2!

**Comment by tk0miya** [comment]
I feel serious bug #3688 has found on 1.6b2. So I'd like to package next beta: 1.6b3 soon.
Is there any problem? If none, I'll do it tomorrow or next day.

**Comment by jfbu** [comment]
+1 for 1.6b3

**Comment by tk0miya** [comment]
Sorry for very late response. I could not take a time for release until now. So I will do that tomorrow. We have monthly hack-a-thon for Sphinx in Japan :-)

**Comment by tk0miya** [comment]
Now I'm just released 1.6b3!
https://pypi.python.org/pypi/Sphinx/1.6b3

**Comment by shimizukawa** [comment]
Topics of Sphinx-1.6 release

```
* Added: entry point; Discovery of builders
* Added: Support remote images and data URI images on non-HTML builders.
* Added: Extension; ``sphinx.ext.imgconverter``. It converts images in the document to appropriate format for builders
* Added: Event; `env-check-consistency` to check consistency to extensions
* Added: Interface; `Domain.check_consistency()` to check consistency
* Added: Options for directives.
* Added: Warnings (non-referenced footnotes and citations, over dedent  on ``literalinclude``, required EPUB3 metadata)
* Added: ``LATEXMKOPTS`` variable for the Makefile
* Experimental: HTML5 writer if ``html_experimental_html5_writer`` is True and docutils 0.13 or later is installed.
* Experimental: HTML5 based epub output if ``html_experimental_html5_writer`` available.
* Changed: Replace the custom smartypants code with Docutils' smart_quotes.
* Changed: LaTeX builder and depended packages
* Changed: ``latexmk`` is required for ``make latexpdf`` on Unix-like platforms.
* Updated: latex output is tested with Ubuntu trusty's texlive packages (Feb. 2014),
* Deprecated: Class ``sphinx.util.compat.Directive``, use ``docutils.parsers.rsr.Directive``
* Deprecated: Variable ``sphinx.util.compat.docutils_version``
* Deprecated: API ``Sphinx.warn()``, ``Sphinx.info()`` and other logging APIs. use ``sphinx.util.logging``
* Deprecated: API ``Sphinx.status_iterator()`` and ``Sphinx.old_status_iterator()``. use ``sphinx.util:status_iterator()``
* Deprecated: API ``Sphinx._directive_helper()``. use ``sphinx.util.docutils.directive_helper()``
* Deprecated: methods of BuildEnvironment; ``set_warnfunc()``, ``note_toctree()``,
  ``get_toc_for()``, ``get_toctree_for()``, ``create_index()``.
* Deprecated: method ``builder.css_files``. use ``app.add_stylesheet()``
* Deprecated: builder ``epub2``
* Deprecated: entry point ``sphinx_themes``. use ``sphinx.html_themes``
* Deprecated: package: ``sphinx.websupport``. use ``sphinxcontrib-websupport``
* Deprecated: ``notice`` LaTeX environment name. use ``sphinxadmonition``
* Removed: Configuration variables ``epub3_contributor``, ``epub3_description``, ``epub3_page_progression_direction``, ``html_translator_class``, ``html_use_modindex``, ``latex_font_size``, ``latex_paper_size``, ``latex_preamble``, ``latex_use_modindex``, ``latex_use_parts``, ``latex_keep_old_macro_names``.
* Removed: node ``termsep``.
* Removed: template ``defindex.html``.
* Removed: LDML format support in ``today``, ```today_fmt`` and ``html_last_updated_fmt``.
* Removed: ``:inline:`` option for the directives of ``sphinx.ext.graphviz`` extension
* Removed: extension ``sphinx.ext.pngmath``
* Removed: function ``sphinx.util.compat.make_admonition()``
```

**Comment by jfbu** [comment]
Also:
```
* Added: LaTeX; Use templates to render tables (refs #3389, 2a37b0e)
```
(It was missing at 1.6 beta1 from CHANGES, I just added it).

There are also some removal of LaTeX packages:
```
* Removed: usage of LaTeX package ``footnote``, replaced by shipped ``footnotehyper-sphinx``
* Removed: usage of LaTeX package ``eqparbox``
* Removed: usage of LaTeX package ``multirow``
* Removed: usage of LaTeX package ``threeparttable``
```
but was this supposedly already included in this @shimizukawa ?
```
* Changed: LaTeX builder and depended packages
```
(there is typo: dependent I think; but I am not sure what this entry refers to)
LaTeX also has some other deprecations.

**Comment by tk0miya** [comment]
Hi all,

As a quick look, there are no blocker for final release. Do you have any blockers or worries for release? If none, I will release 1.6-final at this weekend!
https://github.com/sphinx-doc/sphinx/milestone/23

Note: remaining issues and comments
- #3734 (circleci): This is an issue for developers. So this is not related with the release. It might be merged after final.
- #3736 (Refactor test_latex_table): This is an issue for testing. So this is also not related.
- #3737 (Update messages): This is related with release. This also related with translations. So I will update and merge this ASAP.

**Comment by jfbu** [comment]
I have no blockers for release. I will (most certainly) be away of computers starting Saturday and up to possibly Thursday. So, in case of avalanche of LaTeX bug reports on Monday, if I don't react it is not due to hiding, but to not having internet... hiding will be for later, perhaps ;-)

**Comment by tk0miya** [comment]
Ah, I'd like to ask you about 1.5.6. Now we have some unreleased bug fixes on stable branch.
Is it worthy to release 1.5.6 and 1.6 at same time?

Pros:
* New major release might contain new bugs because 1.6 is much modified and much improved. And it has many incompatible changes also. So providing bug fix release is worthy. It is surely stable. 

Cons:
* With our policy, releasing 1.6 means the EOL of 1.5 series. So it becomes unsupported version at same time as release.


IMO, releasing 1.5.6 is a bit worthy.

**Comment by jfbu** [comment]
I think it is worthwile to release 1.5.6. It does fix some bugs of 1.5.5 and will be stable. One can imagine that release of 1.6 will trigger wave of testing, perhaps revealing an old bug of 1.5.x series. So there is some pro into delaying 1.5.6 to a little after 1.6. However that does not seem to be our policy.

**Comment by jfbu** [comment]
What I said about testing of 1.6 by general Sphinx users might reveal bug of 1.5.x series is not very sensical. So I remove my comment about postponing a bit 1.5.6. As you said some incompatible changes come with 1.6, so some people perhaps will prefer stay with 1.5.x series for a while. So I think it is good to release 1.5.6 which has some bug fixes; with understanding indeed it is released and not supported at same time ...

**Comment by jfbu** [comment]
Hmm.. rather than big discourse I will just say +0 for 1.5.6... no strong opinion !

(deleting my two contradictory earlier comments; in the end I agree 1.5.6 has very few bug fixes, and that it makes sense to only release 1.6, and drop specific release of 1.5.6)

**Comment by tk0miya** [comment]
@jfbu Thank you for comment. I will release both 1.5.6 and 1.6 at this weekend (maybe, in sunday night in JST)

**Comment by jfbu** [comment]
Hi I will be offline until end of next week
(I will possibly be still online until tomorrow 10AM EST)
Wishing good success for 1.6 release !

**Comment by tk0miya** [comment]
I'd forgotten to mention that. I see, I'll try to cover your absent with all my effort!
Have a good offline life :-)

**Comment by tk0miya** [comment]
Now I get a new pull request for 1.6, #3718.

At first, I'm going to release 1.5.6. After that, I will take a look the PR.

**Comment by tk0miya** [comment]
I just released 1.5.6 as the final version of 1.5.x series!

BTW, about the review of #3718 and 1.6 release, I will do them at tomorrow or the next day.
(Sorry, I could not take enough time for releasing...)

**Comment by tk0miya** [comment]
Finally, I just released Sphinx-1.6.1.
From now on, stable branch is used to maintain 1.6.x series.

Note: Because of PyPI's restriction, 1.6 becomes missing version. I'd uploaded wrong package versioned 1.6 by mistake. I just removed it, but It causes to prevent uploading new package as same filename again. sorry for my mistake.

**Comment by tk0miya** [comment]
1.6 contains many many changes:
* 23 incompatible changes
* 7 features removed
* 33 features added
* 29 bug fixed
* 15 deprecated

I think it is much improved since last major release.

Thank you for all your effort.

**Comment by stephenfin** [comment]
> Finally, I just released Sphinx-1.6.1.
> From now on, stable branch is used to maintain 1.6.x series.
> 
> Note: Because of PyPI's restriction, 1.6 becomes missing version. I'd uploaded wrong package versioned 1.6 by mistake. I just removed it, but It causes to prevent uploading new package as same filename again. sorry for my mistake.

fwiw, you can upload wheels and egg files using the same version - just not `tar.gz` files

**Comment by tk0miya** [comment]
Thank you for advice. But I uploaded both wheel and tarball....
